### ISSUE TITLE
Find a better hashing solution

### ISSUE THREAD

**Issue by Mstrodl** [discordapp/module-updater#5]
Sorry, but it seems to mean making a hash of all the files' contents concatenated is generally a bad idea since they aren't guaranteed to be in the same order each time, perhaps you could create a hash for each file instead?

**Comment by SEAPUNK** [comment]
The files get hashed in array order. The hashing is not concurrent, so there is no chance of race condition.

**Comment by Mstrodl** [comment]
@SEAPUNK I think you're missing the point, the issue is that you aren't necessarily always going to get the same order when you call `fs.readdir`

**Comment by SEAPUNK** [comment]
In practice, it's very efficient, and provided you understand its implications i.e. ensuring the order you hashed files in is reflected in the module manifest's files array, verification will always work as expected.

**Comment by Mstrodl** [comment]
@SEAPUNK well then at the very least you can realise that it'd be a bit faster if you could do it all concurrently?

**Comment by jhgg** [comment]
In practice, you could just read the files concurrently and feed them in order to the hasher. JS is single-threaded after all. 

The validation process cares about the module as a whole being valid. The validity of each individual file is kinda irrelevant, as the "repair" process nukes the entire module and re-installs it anyways.

**Comment by SEAPUNK** [comment]
The performance benefit is negligible. Verifying our most file-intensive module (151 files) takes a mere 76ms, so we prioritize minimizing overall module size.

**Comment by jhgg** [comment]
I think there is a way to trick the validity checker into thinking that modules are valid though... assume we have a module with 2 files, "a.txt" and "b.txt"

## a.txt
```
hello there
```

## b.txt
```
my name is jake
```

And we changed them to the following:

## a.txt
```
hello there my name 
```

## b.txt
```
is jake
```

They would probably hash to the same value... 

I mean, there's 2 ways we can work around this.

a) Hash each file individually, then hash those hashes (in order) to form the module hash. (and hey, we can exploit concurrency a bit faster, to the OP's point)
b) Build the hash of each file along with the file name.

## a
```py
def hash_file(filename):
    return hashlib.md5(open(filename).read()).hexdigest()

manifest_hash = hashlib.md5(''.join(hash_file(filename) for filename in filenames]).hexdigest()
```

## b
```py
hasher = hashlib.md5()
for filename in filenames:
    hasher.update(filename)
    hasher.update(open(filename).read())

manifest_hash = hasher.hexdigest()
```

**Comment by SEAPUNK** [comment]
@jhgg Yeah, you're right. Validator can be tricked in this way, however unlikely that case may be. At this rate, we're just cleaning up extreme edge cases, but I don't think changing the generation/validation algorithm to this will take much work.

If I have time after I finish the things I need to first, I'll revisit this and open up a PR for it.

**Comment by aveao** [comment]
Hey, any updates on this?

**Comment by SEAPUNK** [comment]
Nope. This project is currently on hold.
